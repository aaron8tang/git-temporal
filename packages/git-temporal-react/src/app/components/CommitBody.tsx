import React from 'react';
import { style } from 'app/styles';

export interface CommitBodyProps {
  text: string;
  style?: object;
  isExpanded?: boolean;
}

const containerStyle = {
  _extends: ['normalText', 'block'],
  position: 'relative',
  marginLeft: 10,
};
const constrainedStyle = {
  maxHeight: 70,
  overflow: 'hidden',
  marginBottom: 10,
  boxShadow: 'rgba(245, 245, 245, 0.5) 0px -34px 11px -18px inset',
};
const constrainedIndicatorStyle = {
  position: 'absolute',
  bottom: 0,
  display: 'block',
  height: 20,
  width: '100%',
  /* W3C, IE10+, FF16+, Chrome26+, Opera12+, Safari7+ */
  background: 'linear-gradient(#ffffff00, white)',
};

const breakLineStyle = {
  display: 'block',
  marginBottom: 5,
};

const renderBreakLine = (line, index) => {
  return (
    <div key={index} style={style(breakLineStyle)}>
      {line}
    </div>
  );
};

const renderBrText = text => {
  return text.split(/\<br\>/gi).map(renderBreakLine);
};

const renderContstrainedIndicator = isExpanded => {
  return isExpanded ? null : <div style={style(constrainedIndicatorStyle)} />;
};

// numbers for humans have commas
export const CommitBody = (props: CommitBodyProps): JSX.Element => {
  if (!props.text || props.text.replace(/\s/g, '') === '') {
    return null;
  }
  const styles = [containerStyle, props.style];
  if (!props.isExpanded) {
    styles.push(constrainedStyle);
  }
  return (
    <div style={style(styles)}>
      {renderBrText(props.text)}
      {renderContstrainedIndicator(props.isExpanded)}
    </div>
  );
};
