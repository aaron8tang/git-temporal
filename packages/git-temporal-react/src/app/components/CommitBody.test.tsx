import React from 'react';
import { shallow } from 'enzyme';

import { CommitBody } from './CommitBody';

const testText =
  'each<br><br>line<br><br>should<br><br>be<br><br>in a span<br><br>';

describe('components/CommitBody', () => {
  describe('when rendered with text containing <br>s', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallow(<CommitBody text={testText} />);
    });

    test('it should match snapshot (it should have rendered a block element for each line with <br><br>")', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
