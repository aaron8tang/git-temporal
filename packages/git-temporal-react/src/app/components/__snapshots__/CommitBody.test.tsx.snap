// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CommitBody when rendered epoch time 1539369263 (Fri, 12 Oct 2018 18:34:23 GMT) it should match snapshot (it should have rendered a block element for each line with <br><br>") 1`] = `
<div
  style={
    Object {
      "boxShadow": "rgba(245, 245, 245, 0.5) 0px -34px 11px -18px inset",
      "color": "#333333",
      "display": "block",
      "fontSize": 13,
      "fontWeight": "normal",
      "marginBottom": 10,
      "marginLeft": 10,
      "maxHeight": 70,
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <div
    key="0"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  >
    each
  </div>
  <div
    key="1"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  />
  <div
    key="2"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  >
    line
  </div>
  <div
    key="3"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  />
  <div
    key="4"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  >
    should
  </div>
  <div
    key="5"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  />
  <div
    key="6"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  >
    be
  </div>
  <div
    key="7"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  />
  <div
    key="8"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  >
    in a span
  </div>
  <div
    key="9"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  />
  <div
    key="10"
    style={
      Object {
        "display": "block",
        "marginBottom": 5,
      }
    }
  />
  <div
    style={
      Object {
        "background": "linear-gradient(#ffffff00, white)",
        "bottom": 0,
        "display": "block",
        "height": 20,
        "position": "absolute",
        "width": "100%",
      }
    }
  />
</div>
`;
